zndraw/server/events.py

```diff
@@ -148,259 +171,126 @@ def modifier_run(data):
-        except KeyError:
-            return "No host found."
+    emit("scene:set", data["index"], include_self=False, to=_webclients_room(data))
 
 
 @io.on("scene:step")
 def scene_step(data):
-    if "sid" in data:
-        return call("scene:step", to=data["sid"])
-    else:
-        try:
-            # emit to all webclients in the group, if no sid is provided
-            return call("scene:step", to=app.config["ROOM_HOSTS"][session["token"]][0])
-        except KeyError:
-            return "No host found."
+    return call("scene:step", to=_webclients_room(data))
 
 
 @io.on("atoms:download")
 def atoms_download(data):
-    if "sid" in data:
-        sid = data.pop("sid")
-        return call("atoms:download", data["indices"], to=sid)
-    else:
-        try:
-            return call(
-                "atoms:download",
-                data["indices"],
-                to=app.config["ROOM_HOSTS"][session["token"]][0],
-            )
-        except KeyError:
-            return "No host found."
+    return call("atoms:download", data["indices"], to=_webclients_default(data))
 
 
 @io.on("atoms:upload")
 def atoms_upload(data: dict):
-    if "sid" in data:
-        # if the data is sent from the default pyclient, it will have a sid
-        sid = data.pop("sid")
-        emit("atoms:upload", data, include_self=False, to=sid)
-    else:
-        emit("atoms:upload", data, include_self=False, to=session["token"])
+    emit("atoms:upload", data, include_self=False, to=_webclients_room(data))
 
 
 @io.on("atoms:delete")
 def atoms_delete(data: dict):
-    if "sid" in data:
-        # if the data is sent from the default pyclient, it will have a sid
-        sid = data.pop("sid")
-        emit("atoms:delete", data["index"], include_self=False, to=sid)
-    else:
-        emit("atoms:delete", data["index"], include_self=False, to=session["token"])
+    emit("atoms:delete", data["index"], include_self=False, to=_webclients_room(data))
 
 
 @io.on("atoms:length")
 def atoms_length(data: dict):
-    log.debug(f"atoms:length for {data}")
-    if "sid" in data:
-        return call("atoms:length", to=data["sid"])
-    else:
-        try:
-            return call(
-                "atoms:length", to=app.config["ROOM_HOSTS"][session["token"]][0]
-            )
-        except KeyError:
-            return "No host found."
+    return call("atoms:length", to=_webclients_default(data))
 
 
 @io.on("analysis:schema")
 def analysis_schema(data: dict):
     if "sid" in data:
         emit("analysis:schema", data["schema"], include_self=False, to=data["sid"])
     else:
-        try:
-            # emit to all webclients in the group, if no sid is provided
-            emit(
-                "analysis:schema",
-                data["schema"],
-                include_self=False,
-                to=session["token"],
-            )
-        except KeyError:
-            return "No host found."
+        raise ValueError
```

Comment:
Could this cause someone to inadvertently crash the web interface for everyone?

Reply:
The server is quite robust in handling exceptions. If someone uses the UI or the Python interface as intended, this should not be triggered. If someone manually sends messages on the sockets, this might be possible, but there are currently many ways of `killing` the server.